### PR TITLE
Add address to favorites

### PR DESCRIPTION
**Pull request by wazolab**

### CODE REVIEW

File: vidos-config-cypress.json [hunk 1]
```diff
@@ -6,6 +6,7 @@
     ],
     "API_SEARCH": "https://search-dev.teritorio.xyz/search",
     "API_SEARCH_ADDR": "https://api-adresse.data.gouv.fr/search",
+    "API_REVERSE_ADDR": "https://api-adresse.data.gouv.fr/reverse",
```
Comment (frodrigo) [inline vidos-config-cypress.json:9]: Should be refactored as 
```suggestion
    "API_ADDR": "https://api-adresse.data.gouv.fr",
```

**Comment by wazolab** [comment]
Maybe we should make PDF and CSV exports compliant with addresses ?

PDF link example: https://print-dev.teritorio.xyz/v0.1/dev/tourism/pois/favorites.pdf?ids=934332,addr-spb40ppmk
CSV link example: https://dev.appcarto.teritorio.xyz/content/api.teritorio/geodata/v0.1/dev/tourism/pois.csv?ids=934332,addr-spb40ppmk

File: components/Home/Home.vue [hunk 2]
```diff
@@ -300,14 +344,52 @@ async function fetchFavorites(ids: number[]) {
...
+  const promises = Array.from(favoriteAddresses.value.values()).map(async (hash) => {
+    const coordinates = decodeBase32(hash)
+    return await useFetch<ApiSearchResult<ApiAddrSearchResult>>(`${config!.API_ADDR}/reverse?lon=${coordinates.lng}&lat=${coordinates.lat}`)
+  })
+
+  const responses = await Promise.all(promises)
+  // TODO: Handle error or empty data
+  return responses
+    .filter(({ data, error }) => !error.value && data.value?.features.length)
+    .map(({ data }) => ({
```
Comment (wazolab) [inline components/Home/Home.vue:368]: Need to extract this to a function

File: components/Home/Home.vue [hunk 3]
```diff
@@ -71,6 +73,67 @@ const device = useDevice()
...
+        })
+
+      Promise.all(addressHashes.map(async (hash) => {
+        const coordinates = decodeBase32(hash)
+        const address = await $fetch<ApiSearchResult<ApiAddrSearchResult>>(`${config!.API_ADDR}/reverse?lon=${coordinates.lng}&lat=${coordinates.lat}`)
+
+        if (!address || !address.features.length)
+          throw createError({ statusCode: 404, message: `Error while reverse geocoding: ${hash}` })
+
+        favoritesStore.toggleFavoriteAddr({
```
Comment (wazolab) [inline components/Home/Home.vue:102]: Need to extract this to a function

File: components/Home/Home.vue [hunk 4]
```diff
@@ -270,8 +301,18 @@ watch(mode, () => {
   routerPushUrl(hash)
 })
 
-watch(isModeFavorites, async () => {
-  await handleFavorites()
+watch(isModeFavorites, async (isEnabled) => {
+  if (isEnabled) {
+    if (favoriteCount.value !== favoriteFeatures.value.length
```
Comment (wazolab) [inline components/Home/Home.vue:306]: Could make a computed if this

File: components/Home/Home.vue [hunk 5]
```diff
@@ -354,6 +436,16 @@ function onToggleFavoritesMode() {
   }
 }
 
+async function toggleNoteBookMode() {
+  if (favoriteCount.value !== favoriteFeatures.value.length
```
Comment (wazolab) [inline components/Home/Home.vue:440]: Could make a computed if this

File: components/Home/Home.vue [hunk 6]
```diff
@@ -71,6 +73,66 @@ const device = useDevice()
...
+    try {
+      const addressHashes: string[] = []
+
+      favs
+        .split(',')
+        .forEach((id) => {
+          if (!Number.isNaN(Number(id)))
+            favoritesStore.toggleFavorite(Number(id))
+
+          if (id.startsWith('addr-'))
```
Comment (frodrigo) [inline components/Home/Home.vue:87]: Can you use `addr:` as prefix. To homogeneous with `ref:` and other `:` was have in the API.

File: components/Home/Home.vue [hunk 7]
```diff
@@ -71,6 +73,66 @@ const device = useDevice()
...
+            addressHashes.push(id.substring(5))
+        })
+
+      await Promise.all(addressHashes.map(async (hash) => {
+        const coordinates = decodeBase32(hash)
+        const address = await $fetch<ApiSearchResult<ApiAddrSearchResult>>(`${config!.API_ADDR}/reverse?lon=${coordinates.lng}&lat=${coordinates.lat}`)
+
+        if (!address || !address.features.length)
+          throw createError({ statusCode: 404, message: `Error while reverse geocoding: ${hash}` })
```
Comment (frodrigo) [inline components/Home/Home.vue:96]: This exception stop all ?
If we cannot do the reverse, we can continue anyway.

File: components/Home/Home.vue [hunk 8]
```diff
@@ -71,6 +73,66 @@ const device = useDevice()
...
+          geometry: address.features[0].geometry,
+          properties: {
+            internalType: 'address',
+            metadata: {
+              id: address.features[0].properties.id as ApiPoiId,
+            },
+            name: address.features[0].properties.label,
+            vido_zoom: address.features[0].properties.type === 'municipality'
+              ? MAP_ZOOM.selectionZoom.municipality
+              : MAP_ZOOM.selectionZoom.streetNumber,
```
Comment (frodrigo) [inline components/Home/Home.vue]: Were are not doing a geocoding. We should use a detailled zoom level in all case.

File: components/Home/Home.vue [hunk 9]
```diff
@@ -300,14 +343,52 @@ async function fetchFavorites(ids: number[]) {
...
+    return []
+
+  const promises = Array.from(favoriteAddresses.value.values()).map(async (hash) => {
+    const coordinates = decodeBase32(hash)
+    return await useFetch<ApiSearchResult<ApiAddrSearchResult>>(`${config!.API_ADDR}/reverse?lon=${coordinates.lng}&lat=${coordinates.lat}`)
+  })
+
+  const responses = await Promise.all(promises)
+  // TODO: Handle error or empty data
+  return responses
```
Comment (frodrigo) [inline components/Home/Home.vue:341]: Duplicate code. Same code on mount ?

File: components/Home/Home.vue (hunk 7 above)
Comment (wazolab) [reply components/Home/Home.vue:96]: mmh I think it just handles the error and throws it but it's non-fatal.

File: components/Home/Home.vue (hunk 9 above)
Comment (wazolab) [reply components/Home/Home.vue:341]: This part is duplicated ... I could make a function of it eventually.

```js
const coordinates = decodeBase32(hash)
return await useFetch<ApiSearchResult<ApiAddrSearchResult>>(`${config!.API_ADDR}/reverse?lon=${coordinates.lng}&lat=${coordinates.lat}`)
```

File: components/Home/Home.vue [hunk 10]
```diff
@@ -71,6 +73,66 @@ const device = useDevice()
...
+              id: address.features[0].properties.id as ApiPoiId,
+            },
+            name: address.features[0].properties.label,
+            vido_zoom: address.features[0].properties.type === 'municipality'
+              ? MAP_ZOOM.selectionZoom.municipality
+              : MAP_ZOOM.selectionZoom.streetNumber,
+            display: {
+              icon: address.features[0].properties.type === 'municipality'
+                ? 'city'
+                : 'map-marker-alt',
```
[the remark this answers is not in the document]
Comment (wazolab) [reply components/Home/Home.vue]: I think it's better to keep it this way in order to ensure consistency between popups.

If we set all the favorite address POI with 'map-marker-alt' we will have such cases during user journey:

After searching a city:
Screenshot from 2024-05-22 21-55-37

After toggeling favorite mode:
Screenshot from 2024-05-22 21-55-53

File: components/Home/Home.vue (hunk 8 above)
Comment (wazolab) [reply components/Home/Home.vue]: Yep in favorite mode vido_zoom isn't even used I think.

[InitialBbox](https://github.com/teritorio/vido/blob/ebc5e01b71c0ab975b9415d4891882a4f6181cb7/components/Home/Home.vue#L289) is managing the zoom in this case :
```js
initialBbox.value = getBBoxFeatures(favoriteFeatures.value)
```

File: components/Home/Home.vue (hunk 10 above)
[the remark this answers is not in the document]
Comment (wazolab) [reply components/Home/Home.vue]: I think the reverse API call gives me a city in return, first feature is the city, second a housnumber close to the coords I provided :
image

Should we keep it as is ? Or I remove the city icon support ?

[the remark this answers is not in the document]
Comment (frodrigo) [reply components/Home/Home.vue]: Ok, keep it. You can add `&limit=1` to reverse query.

It works, but is almost a bug on geocoder side. There is no garantee that will always works, especially in time.

**Pull request merged**